pkg/networkservice/common/updatepath/common.go

```diff
@@ -17,74 +17,77 @@
 package updatepath
 
 import (
-	"context"
-
-	"github.com/golang/protobuf/ptypes"
+	"github.com/google/uuid"
 	"github.com/networkservicemesh/api/pkg/api/networkservice"
 	"github.com/pkg/errors"
-	"google.golang.org/grpc/credentials"
-	"google.golang.org/grpc/peer"
-
-	"github.com/networkservicemesh/sdk/pkg/tools/token"
 )
 
-type commonUpdatePath struct {
-	name           string
-	tokenGenerator token.GeneratorFunc
-}
+/*
+	Logic for Update path:
 
-func (u *commonUpdatePath) updatePath(ctx context.Context, conn *networkservice.Connection) error {
-	// We need a real path if we don't have one
+	1. If current path segment.Name is equal to segmentName passed, it will just update current connection.Id and exit.
+	2. If current path segment.Name is not equal to segmentName:
+		2.1 if path has next segment available and next name is segmentName, take Id from next path segment
+		2.2 if no next path segment available, it will add one more path segment and generate new Id, update connection.Id
+		2.3 if path has next segment available, but next name is not equal to segmentName, will return error.
+	3. if Index == 0, and there is no current segment will add one.
+*/
+func updatePath(conn *networkservice.Connection, segmentName string) (*networkservice.Connection, error) {
 	if conn == nil {
-		return errors.New("NetworkServiceRequest.Connection must not be nil")
+		conn = &networkservice.Connection{}
 	}
 
 	// If we don't have a Path, add one
-	if conn.GetPath() == nil {
-		conn.Path = &networkservice.Path{}
+	if conn.GetPath() == nil || len(conn.GetPath().GetPathSegments()) == 0 {
+		if conn.Path == nil {
+			conn.Path = &networkservice.Path{}
+		}
+		conn.Path.Index = 0
+		// Replace path for first segment.
+		conn.Path.PathSegments = []*networkservice.PathSegment{
+			{
+				Name: segmentName,
+				Id:   conn.Id,
```

Comment:
What if conn.Id hasn't been set elsewhere?  For example for a leaf client... this is taking conn.Id from the request... which is fine in cases like xconnect or nsmgr chains which chain a client from a server... but not correct for general clients that may not have one set at all.

Reply:
OK... just read down the server code and see it too is simply passing in the request.GetConnection() to update path... the question is... where are ids being generated?

Reply:
OK... just found id generation for conn.Id below for the nextIndex >= case... my underlying point here is: in the starting a new path case, if conn.Id == "" we should generate one.

Reply:
ok, fixed.